wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/jpa/JpaUserDevfileDao.java

```diff
@@ -0,0 +1,295 @@
+
+import com.google.common.annotations.Beta;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.persist.Transactional;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.persistence.EntityManager;
+import javax.persistence.TypedQuery;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.Page;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.workspace.devfile.UserDevfile;
+import org.eclipse.che.api.devfile.server.model.impl.UserDevfileImpl;
+import org.eclipse.che.api.devfile.server.spi.UserDevfileDao;
+import org.eclipse.che.commons.lang.Pair;
+import org.eclipse.che.core.db.jpa.DuplicateKeyException;
+import org.eclipse.che.core.db.jpa.IntegrityConstraintViolationException;
+
+/** JPA based implementation of {@link UserDevfileDao}. */
+@Singleton
+@Beta
+public class JpaUserDevfileDao implements UserDevfileDao {
+
+  @Inject private Provider<EntityManager> managerProvider;
+  public static final List<Pair<String, String>> DEFAULT_ORDER =
+      ImmutableList.of(new Pair<>("id", "ASC"));
+
+  @Override
+  public UserDevfile create(UserDevfile devfile) throws ConflictException, ServerException {
+    requireNonNull(devfile);
+    try {
+      doCreate(new UserDevfileImpl(devfile));
+    } catch (DuplicateKeyException ex) {
+      throw new ConflictException(
+          format("Devfile with name '%s' already exists for current user", devfile.getName()));
+    } catch (IntegrityConstraintViolationException ex) {
+      throw new ConflictException(
+          "Could not create devfile with creator that refers on non-existent user");
+    } catch (RuntimeException ex) {
+      throw new ServerException(ex.getMessage(), ex);
+    }
+    return new UserDevfileImpl(devfile);
+  }
+
+  @Override
+  public Optional<UserDevfile> update(UserDevfile update)
+      throws ConflictException, ServerException {
+    requireNonNull(update);
+    try {
+      return doUpdate(new UserDevfileImpl(update)).map(UserDevfileImpl::new);
+    } catch (DuplicateKeyException ex) {
+      throw new ConflictException(
+          format("Devfile with name '%s' already exists for current user", update.getName()));
+    } catch (RuntimeException ex) {
+      throw new ServerException(ex.getLocalizedMessage(), ex);
+    }
+  }
+
+  @Override
+  public void remove(String id) throws ServerException {
+    requireNonNull(id);
+    try {
+      doRemove(id);
+    } catch (RuntimeException ex) {
+      throw new ServerException(ex.getLocalizedMessage(), ex);
+    }
+  }
+
+  @Override
+  @Transactional(rollbackOn = {ServerException.class})
```

Comment:
Since `requireNonNull` throws runtime exception it makes sense to add it to rollbackOn list.